content/advent-2015/gopy.md

```diff
@@ -0,0 +1,342 @@
+        
+        return Py_BuildValue("O&", cgopy_cnv_c2py_string, &c_gopy_ret);
+}
+
+/* functions for package hello */
+static PyMethodDef cpy_hello_methods[] = {
+        {"Hello", cpy_func_hello_Hello, METH_VARARGS, "Hello(str name) str\n\nHello greets someone.\n"},
+        {NULL, NULL, 0, NULL}        /* Sentinel */
+};
+
+PyMODINIT_FUNC
+inithello(void)
+{
+        PyObject *module = NULL;
+        
+        /* make sure Cgo is loaded and initialized */
+        cgo_pkg_hello_init();
+        
+        module = Py_InitModule3("hello", cpy_hello_methods, "hello is a simple package\n");
+        
+}
+```
+
+The generated `C` code uses the `CPython` API to create a `CPython` module, fill
+the _ad_ _hoc_ structures and connect the `cgo` functions with the `CPython`
+infrastructure, together with the `python` docstrings.
+
+## Implementation notes
+
+`gopy`, like [gomobile](https://golang.org/x/mobile/cmd/gomobile), has the
+interesting task of binding two languages together, each of them with a garbage
+collector.
+Thus, `gopy` needs to arrange somehow for the bindings to setup a kind of
+protocol so that each garbage collector knows when a foreign value is not needed
+anymore.
+To that end, `gopy` notes when a `Go` value crosses the `python` boundary and
+increments a counter so the `Go` garbage collector will not try to collect it
+while the `python` side has a hold on it.
+On the `python` side, the `C` extension module is generated in such a way that
+the `python` garbage collector will tell the `Go` side when that value is not
+needed anymore, as far as the `python` side is concerned.
+ 
+As a consequence, any `Go` value that somehow crosses the language boundary
+needs to be allocated on the heap: `gopy` takes care of that while generating
+the `cgo` package.
+
+Schematically, a call from `python` to a `Go` function looks like:
+
+```
+// Call sequence (lots of hand-waving)
+
+-> python:    hello.Hello("foo")
+ -> cpython:  cpy_func_hello_Hello(...)
+  -> cgo:     cgo_func_hello_Hello(...)
+   -> go:     hello.Hello(...)
+```
+
+so, while `gopy` will help bring some of the `Go` goodies to `python`, the
+wrapped entities should not be called inside a very tight loop.
+
+Another interesting point to note is that, as `gopy` uses `cgo`, it is subject
+to the new and more stringent rules that are devised for [Go-1.6](https://github.com/golang/proposal/blob/master/design/12416-cgo-pointers.md).
+As a consequence, some internal parts of `gopy` will need to be modified to
+generate code that follows these news.
+Unfortunately, this probably means another slight performance hit with regard to
+what could be achieved with `Go-1.5`.
+
+`gopy` is currently able to wrap `struct` types, named types, slice types and array
+types, as well as the methods which are attached to it.
+`gopy` is also clever enough to translate the [comma-ok idiom](https://golang.org/doc/effective_go.html) into its `python` equivalent (_ie:_ raising an `Exception`.)
```

Comment:
comma-ok is used in various places. map look ups, type assertions. do you just mean for map look ups or in all the places?